server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1302,78 +1396,127 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer,
 
-        if (quantity > 0) {
+        log.debug("Locking pools: {}", poolQuantityMap.keySet());
+
+        List<Pool> pools = poolCurator.lockAndLoad(poolQuantityMap.keySet());
+
+        for (Pool pool : pools) {
+            log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
+        }
+
+        Map<String, PoolQuantity> poolQuantities = new HashMap<String, PoolQuantity>();
+        boolean quantityFound = false;
+        for (Pool pool : pools) {
+            Integer quantity = poolQuantityMap.get(pool.getId());
+            if (quantity > 0) {
+                quantityFound = true;
+            }
+            poolQuantities.put(pool.getId(), new PoolQuantity(pool, quantity));
+            poolQuantityMap.remove(pool.getId());
+        }
+
+        if (!poolQuantityMap.isEmpty()) {
+            throw new IllegalArgumentException(i18n.tr("Subscription pool(s) {0} do not exist.",
+                    poolQuantityMap.keySet()));
+        }
+        if (quantityFound) {
             log.info("Running pre-entitlement rules.");
             // XXX preEntitlement is run twice for new entitlement creation
-            ValidationResult result = enforcer.preEntitlement(consumer, pool, quantity, caller);
-
-            if (!result.isSuccessful()) {
-                log.warn("Entitlement not granted: {}", result.getErrors().toString());
-                throw new EntitlementRefusedException(result);
+            Map<String, ValidationResult> results = enforcer.preEntitlement(consumer,
+                    poolQuantities.values(), caller);
+
+            for (Entry<String, ValidationResult> entry : results.entrySet()) {
+                ValidationResult result = entry.getValue();
+                if (!result.isSuccessful()) {
+                    log.warn("Entitlement not granted: {} for pool: {}",
+                            result.getErrors().toString(), entry.getKey());
+                    throw new EntitlementRefusedException(results);
+                }
             }
         }
+
         EntitlementHandler handler = null;
-        if (entitlement == null) {
+        if (entitlements == null) {
             handler = new NewHandler();
         }
+        else if (entitlements.keySet().size() != poolQuantities.keySet().size() ||
+                !entitlements.keySet().containsAll(poolQuantities.keySet()) ||
+                !poolQuantities.keySet().containsAll(entitlements.keySet())) {
+            throw new IllegalArgumentException(
+                            i18n.tr("Argument mismatch in entitlement update, number of pools: {}," +
+                            " number of entitlements: {}",
+                            entitlements.keySet().size(), poolQuantityMap.keySet().size()));
+        }
         else {
             handler = new UpdateHandler();
         }
 
-        // Grab an exclusive lock on the consumer to prevent deadlock.
-        consumer = consumerCurator.lockAndLoad(consumer);
```

Comment:
@nguyenfilip yup, since that was the only update on the consumer :) much lower risk of lock waits now :) this was actually the incentive for computing compliance asynchronously.